spring-web/src/main/java/org/springframework/http/server/reactive/ReactorServerHttpHeaders.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.http.server.reactive;
+
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import io.netty.handler.codec.http.HttpHeaders;
+
+import org.springframework.lang.Nullable;
+import org.springframework.util.MultiValueMap;
+
+/**
+ * {@code MultiValueMap} implementation for wrapping Reactor Netty HTTP headers.
+ *
+ * @author Brian Clozel
+ * @since 5.1
+ */
+class ReactorServerHttpHeaders implements MultiValueMap<String, String> {
```

Comment:
A few questions on the class name.

1. `HttpHeaders` is from Netty so probably shouldn't be called Reactor? It could be used with other Netty based runtimes too besides Reactor.
2. The Netty `HttpHeaders` are for client and server use, so "Server" is a misnomer.
3. Both Netty and Spring have "HttpHeaders", and this class is an instance of neither, so calling it `HttpHeaders` is not ideal. It is a MultiValueMap but that would be long. It is an Adapter so perhaps `NettyHeadersAdapter`?
4. 

For Jetty the adapted `HttpFields` is also both client and server. Perhaps `JettyHeadersAdapter`. 

For Tomcat and Undertow they're server specific but no strong reason to use "Server" (what else could they be?) so `TomcatHeadersAdapter` and `UndertowHeadersAdapter` for consistency.